Add support for Hornbill Y4 Smart Lock family

jayPare opened the pull request:
Adding support for [ Hornbill Y4 Smart Lock family](https://hornbilllock.com/product/y4-fingerprint-door-lock/).

![image](https://github.com/user-attachments/assets/3018ee50-57af-4e39-8d8e-ede976fd0e61)

Here are the query results:
[query_things_data_model.json](https://github.com/user-attachments/files/19011589/query_things_data_model.json)
[query_device_details.json](https://github.com/user-attachments/files/19011595/query_devide_details.json)

jayPare commented:
This was based on #1830  and #2824, but I still need to understand some parts. Maybe @rudironsonijr and @josefelixh can chip in?

make-all commented on custom_components/tuya_local/devices/hornbill_smart_lock.yaml:
```diff
@@ -0,0 +1,311 @@
+      # TODO: Not sure it belongs in lock entity
+      - id: 18
+        name: open_inside
+        type: bool
```
boolean

jayPare commented:
@make-all I'm trying to run test_device_config.py locally but I always get this:
`ModuleNotFoundError: No module named 'custom_components.tuya_local'`

I'm in a venv and installed all the modules using requirements-dev.txt. I also tried to add `custom_components/` and `custom_components/tuya_local/` in PYTHONPATH without any success.

What am I missing?

make-all commented:
I don't need to add anything to my PYTHONPATH to run pytest. To run the utils with python, I need to add . to PYTHONPATH though.

jayPare commented:
> I don't need to add anything to my PYTHONPATH to run pytest. To run the utils with python, I need to add . to PYTHONPATH though.

So you simply run `python3 tests/test_device_config.py`?

jayPare commented:
> I don't need to add anything to my PYTHONPATH to run pytest. To run the utils with python, I need to add . to PYTHONPATH though.

Adding . seems to get me further, but now I get:

```
Traceback (most recent call last):
  File "/home/jpare/git/tuya-local/tests/test_device_config.py", line 22, in <module>
    from .const import GPPH_HEATER_PAYLOAD, KOGAN_HEATER_PAYLOAD
ImportError: attempted relative import with no known parent package
```

jayPare commented:
> I don't need to add anything to my PYTHONPATH to run pytest. To run the utils with python, I need to add . to PYTHONPATH though.

My bad, I need to use `pytesttests/test_device_config.py`. Still a couple of TODO in hornbill_y4_smart_lock.yaml I need help with.

make-all commented on custom_components/tuya_local/devices/hornbill_y4_smart_lock.yaml:
```diff
@@ -0,0 +1,312 @@
+          - dps_val: closed
+            value: closed
+  - entity: select
+    name: autolock_delay_level
```
Entity name should be human readable: "Autolock delay..." The name also needs to make clear the difference between this setting and "Auto lock time" above.

make-all commented on custom_components/tuya_local/devices/hornbill_y4_smart_lock.yaml:
```diff
@@ -0,0 +1,312 @@
+        range:
+          min: 1
+          max: 1800
+  - entity: select
```
Is this settable? it seems more like a sensor.

make-all commented on custom_components/tuya_local/devices/hornbill_y4_smart_lock.yaml:
```diff
@@ -0,0 +1,312 @@
+        type: boolean
+        name: switch
+  - entity: number
+    name: auto_lock_time
```
Name should be human readable. And time seems to mean "delay" here (not wall clock time).

make-all commented on custom_components/tuya_local/devices/hornbill_y4_smart_lock.yaml:
```diff
@@ -0,0 +1,312 @@
+        type: boolean
+        name: switch
+  - entity: switch
+    name: automatic_lock
```
Name should be human readable

jayPare commented on custom_components/tuya_local/devices/hornbill_y4_smart_lock.yaml:
```diff
@@ -0,0 +1,312 @@
+      # TODO: Why not working? PR#2824 also didn't include it.
+      # - id: 67
+      #   type: string
+      #   name: unlock_offline_pd
```
Done.

jayPare reviewed (commented):
I've addressed most of the comments, but there is still 1 TODO left. Also, to try it, I add th yaml file to `/homeassistant/custom_components/tuya_local/devices/`, but at the configuration, I'm not getting Hornbill Smart Lock in the list of proposed device. Anything else I need to do to test it locally?

jayPare commented:
> I've addressed most of the comments, but there is still 1 TODO left. Also, to try it, I add th yaml file to `/homeassistant/custom_components/tuya_local/devices/`, but at the configuration, I'm not getting Hornbill Smart Lock in the list of proposed device. Anything else I need to do to test it locally?

Adding what I see in HA logs:
```
[33m2025-03-11 15:51:59.879 WARNING (MainThread) [custom_components.tuya_local.config_flow] Adding Bluetooth gateway device with product id yy2bmcoh
[33m2025-03-11 15:52:00.301 WARNING (MainThread) [custom_components.tuya_local.config_flow] Device matches None with quality of 0%. DPS: {"updated_at": 1741722712.6835952, "61": "AAAB"}
[33m2025-03-11 15:52:00.301 WARNING (MainThread) [custom_components.tuya_local.config_flow] Include the previous log messages with any new device request to https://github.com/make-all/tuya-local/issues/
```

jayPare commented:
> > I've addressed most of the comments, but there is still 1 TODO left. Also, to try it, I add th yaml file to `/homeassistant/custom_components/tuya_local/devices/`, but at the configuration, I'm not getting Hornbill Smart Lock in the list of proposed device. Anything else I need to do to test it locally?
> 
> Adding what I see in HA logs:
> 
> ```
> [33m2025-03-11 15:51:59.879 WARNING (MainThread) [custom_components.tuya_local.config_flow] Adding Bluetooth gateway device with product id yy2bmcoh
> [33m2025-03-11 15:52:00.301 WARNING (MainThread) [custom_components.tuya_local.config_flow] Device matches None with quality of 0%. DPS: {"updated_at": 1741722712.6835952, "61": "AAAB"}
> [33m2025-03-11 15:52:00.301 WARNING (MainThread) [custom_components.tuya_local.config_flow] Include the previous log messages with any new device request to https://github.com/make-all/tuya-local/issues/
> ```

Here is a screenshot of the choices I'm getting:
![image](https://github.com/user-attachments/assets/d63a7b77-7fd5-400a-ad16-301fa0ef1d7a)

make-all commented:
I think your types were OK to start with, the error message is auto-generated by Python, so using Python's internal names for the types. But those dps need to be marked `optional: true` because they are missing from what the device is reporting (at least sometimes)

jayPare commented:
Result as of now: 
![image](https://github.com/user-attachments/assets/3f7b262b-7829-4ddb-b9b7-784ab2258ebe)

jayPare commented:
@make-all have you seen some lock where the lock state is reported correctly, but when I hit unlock, the device stays locked since the motor is going in the lock direction. Same goes for when I hit lock, the device stays unlocked since the motor is going in the unlock direction. In Smart Life I have a "Unlock direction" option (left/right), but it is set correctly since when I lock/unlock from Smart Life app, it's working correclty.

I thought maybe the DP 46 needs a lock direction DP to know which side go, but I don't have any DP corresponding to lock direction. There is a DP called reverse_lock, but when I change the  "Unlock direction" option in Smart Life and update device status in Tuya Developer Platform (debug section), it has no effect on the reverse_lock switch. Same for when I trigger reverse_lock from Tuya Developer Platform, the "Unlock direction" option doesn't change in Smart Life.

I don't see any other DP that could match the lock_direction. I even try to see in Tuya Developer Platform if anything was changing when updating "Unlock direction" and Smart Life and I found nothing. Maybe there is another way to have DP46 go in a different direction?

make-all commented:
reverse_lock I think corresponds to a "snib" function, where you can lock the lock in an unlocked (maybe also locked) state.

I don't think you can compare lock/unlock functionality to the app the device is paired with, as the way secure Tuya locks work is the app shares a key with the lock when pairing so that only that app can be used to lock/unlock the lock (it is a separate key from the local_key used for communication).

jayPare commented:
Ok so do you know what could cause the lock to go in opposite direction when locking/unlocking it?

make-all commented on custom_components/tuya_local/devices/hornbill_y4_smart_lock.yaml:
```diff
@@ -0,0 +1,481 @@
+  # lock record reporting. In addition to these two dp, you also need to use
+  # dp47=lock status to report the lock status, so that the unlock/lock status
+  # displayed on the app is consistent with the local hardware.
+  - entity: lock
```
Why is this a separate entity from the lock entity above? The one above does not have a lock dp, so merging this in should be possible.

jayPare replied:
Done.

jayPare commented:
Correct me if I'm wrong, but from what I understood, it's not really possible to add the unlock function to tuya-local. I was mainly adding support for this lock in hope of getting the unlock function available in HA, but it was a good exercise to play with tuya-local and add available support for this lock in tuya-local. I hope with my last patch everything is fine.
